blockdag/ghostdag.go

```diff
@@ -0,0 +1,155 @@
+package blockdag
+
+import "github.com/pkg/errors"
+
+func (dag *BlockDAG) selectedParentAnticone(node *blockNode) (*blockHeap, error) {
+	anticoneSet := newSet()
+	anticoneHeap := newUpHeap()
+	selectedParentPast := newSet()
+	var queue []*blockNode
+	for _, parent := range node.parents {
+		if parent == node.selectedParent {
+			continue
+		}
+		anticoneSet.add(parent)
+		queue = append(queue, parent)
+	}
+	for len(queue) > 0 {
+		var current *blockNode
+		current, queue = queue[0], queue[1:]
+		for _, parent := range current.parents {
+			if anticoneSet.contains(parent) || selectedParentPast.contains(parent) {
+				continue
+			}
+			isAncestorOfSelectedParent, err := dag.isAncestorOf(parent, node.selectedParent)
+			if err != nil {
+				return nil, err
+			}
+			if isAncestorOfSelectedParent {
+				selectedParentPast.add(parent)
+				continue
+			}
+			anticoneSet.add(parent)
+			anticoneHeap.Push(parent)
+			queue = append(queue, parent)
+		}
+	}
+	return &anticoneHeap, nil
+}
+
+// blueAnticoneSize returns the blue anticone size of 'block' from the worldview of 'context'.
+// Expects 'block' to be ∈ blue-set(context)
+func (dag *BlockDAG) blueAnticoneSize(block, context *blockNode) (uint32, error) {
+	for current := context; current != nil; current = current.selectedParent {
```

Comment:
why `current != nil` is the condition to break the loop, shouldn't it be `current == genesis`?

Reply:
It doesn't really matter. It's done only for sanity to prevent nil pointer exception